hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -1,36 +1,87 @@
+    case TableAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case MatrixAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case TableFilter(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case TableMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableAggregateByKey(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case TableKeyByAndAggregate(child, _, _, _, _) => if (i == 1) child.typ.globalEnv.m else if (i == 2) child.typ.rowEnv.m else empty
+    case MatrixMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixFilterRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixMapCols(child, _, _) => if (i == 1) child.typ.colEnv.m else empty
+    case MatrixFilterCols(child, _) => if (i == 1) child.typ.colEnv.m else empty
+    case MatrixMapEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixFilterEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
+    case MatrixAggregateColsByKey(child, _, _) => if (i == 1) child.typ.rowEnv.m else if (i == 2) child.typ.globalEnv.m else empty
+    case MatrixAggregateRowsByKey(child, _, _) => if (i == 1) child.typ.colEnv.m else if (i == 2) child.typ.globalEnv.m else empty
+    case _ => empty
+  }
+}
+
+object AggBindings {
+  private val empty: Array[(String, Type)] = Array()
+
+  def apply(x: BaseIR, i: Int): Iterable[(String, Type)] = x match {
+    case AggLet(name, value, _) => if (i == 1) Array(name -> value.typ) else empty
+    case AggExplode(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case AggArrayPerElement(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case ArrayAgg(a, name, _) => if (i == 1) Array(name -> a.typ.asInstanceOf[TContainer].elementType) else empty
+    case TableAggregate(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixAggregate(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case TableAggregateByKey(child, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case TableKeyByAndAggregate(child, _, _, _, _) => if (i == 1) child.typ.rowEnv.m else empty
+    case MatrixMapRows(child, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixMapCols(child, _, _) => if (i == 1) child.typ.entryEnv.m else empty
+    case MatrixAggregateColsByKey(child, _, _) => if (i == 1) child.typ.entryEnv.m else if (i == 2) child.typ.colEnv.m else empty
+    case MatrixAggregateRowsByKey(child, _, _) => if (i == 1) child.typ.entryEnv.m else if (i == 2) child.typ.rowEnv.m else empty
+    case _ => empty
   }
 }
+
+object TransitiveBindings {
```

Comment:
What's this function doing?  I don't understand the name (or the code).  It is adding the bindings for `ir` at child `i` to `env` and `aggEnv`?  Maybe `AddBindingsToEnv`?

One operation we might want, if we have `ir`, `i`, `env/aggEnv`, where `env/aggEnv` is the environment where `ir` is to be evaluated, then the result is is the environment that appears and child `i`.  Then I think this becomes the simple composition of Bindings/AggBindings and this operation ... `ChildEnv`?

One thing that feels wrong with this is that a relational IR shouldn't have an environment associated with it, so the inputs don't seem quite right.

Reply:
This function computes the env + agg env at a child `i` of a node `ir`, given the previous env and aggEnv.

Reply:
The total separation of relational nodes from the binding stuff was part of the  reason everything had gone to hell. You'll see that they now appear up above in Binds

Reply:
> One operation we might want, if we have ir, i, env/aggEnv, where env/aggEnv is the environment where ir is to be evaluated, then the result is is the environment that appears and child i. Then I think this becomes the simple composition of Bindings/AggBindings and this operation ... ChildEnv?

That's exactly what this function is supposed to be :) The name is terrible...

Reply:
> The total separation of relational nodes from the binding stuff was part of the reason everything had gone to hell.

While I'm disagreeing, appropriate separation is appropriate.  For example, relational nodes do not having environments.  It should not be possible to ask if a variable is bound on, say, a MatrixMapRows.  To that end:

 - I'd like to see Bindings and AggBindings assert the child being queried is a value IR,
 - ChildEnv should have 3 entry points, and the MatrixIR/TableIR entrypoints shouldn't take environment arguments
 - TypeCheck itself should probably have 3 entrypoints in the same way